File: HashTable/hash_table.cpp
#include <iostream>

// Table Size
const int TABLE_SIZE = 128;

// HashNode class
struct HashNode {
    public:
        // Key
        int key;

        // Value
        int value;

        HashNode(int key, int value) : key(key), value(value) {};
};

// HashTable class
class HashTable {
    private:
        // Elements storage
        HashNode** table;

        // Current map size
        int size_;

    public:
        HashTable() {
            size_ = 0;
            table = new HashNode*[TABLE_SIZE];
            // Init buckets
            for (int i = 0; i < TABLE_SIZE; i++) {
                table[i] = nullptr;
            }
        };

        ~HashTable() {
            for (int i = 0; i < TABLE_SIZE; i++) {
                if (table[i] != nullptr)
                    delete table[i];
                delete[] table;
            }
        };

        // Hash Function
        std::size_t hashFunc(int key) {
            return key % TABLE_SIZE;
        };

        // Element insert
        void insert(int key, int value) {
            std::size_t hash = hashFunc(key);
            while (table[hash] != nullptr && table[hash]->key != key) {
                hash = hashFunc(++hash);
            }

            if (table[hash] != nullptr) {
                delete table[hash];
            }

            size_++;
            table[hash] = new HashNode(key, value);
        };

        // Element search
        int get(int key) {
            std::size_t hash = hashFunc(key);
            while (table[hash] != nullptr && table[hash]->key != key) {
                hash = hashFunc(++hash);
            }

            return table[hash] != nullptr ?  table[hash]->value : -1;
        };

        // Is key in map
        bool contains(int key) {
            return get(key) != -1;
        };

        // Display map content
        void display(void) {
            for (int i = 0; i < TABLE_SIZE; i++) {
                if (table[i] != nullptr) {
                    std::cout <<
                        "{K: " << table[i]->key <<
                        " V: " << table[i]->value << 
                        "}" << std::endl;
                }
            }
        };

        // Element delete
        void remove(int key) {
            std::size_t hash = hashFunc(key);
            while (table[hash] != nullptr) {
                if (table[hash]->key != key) {
                    hash = hashFunc(++hash);
                    continue;
                }

                break;
            }

            if (table[hash] == nullptr) 
                return;
            
            size_--;
            delete table[hash];
            table[hash] = nullptr;
        };

        // Size of map
        int size() {
            return size_;
        };
};

File: HashTable/hash_table.js

class HashNode {
  constructor(k, v) {
    this.key = k
    this.value = v
  }
}

class HashTable {
  static TABLE_SIZE = 128

  constructor() {
    this.m = new Array(TABLE_SIZE)
    this.size = 0
  }

  _hashFunction(k) {
    return k % TABLE_SIZE
  }

  getHashFromTable(k) {
    let hash_ = this._hashFunction(k)

    while (this.m[hash_]) {
      if (this.m[hash_].key !== k) {
        hash_ = this._hashFunction(hash_ + 1)
        continue
      }
      
      return hash_
    }

    return null
  }

  insert(k, v) {
    this.size++
    this.m[this.getHashFromTable(k)] = new HashNode(k, v)
  }

  get(k) {
    const hash_ = this.getHashFromTable(k)
    return this.m[hash_] ? this.m[hash_].value : -1
  }
 
  contains(k) {
    return this.get(k) !== -1
  }

  display() {
    this.m
      .filter(node => !!node)
      .forEach(node => console.log(`{K: ${node.key}, V: ${node.value}}`))
  }

  remove(k) {
    const hash_ = this.getHashFromTable(k)

    if (!this.m[hash_]) {
      return
    }

    this.size--
    this.m[hash_] = null
  }
}
